tests/test_fairworkflow.py

```diff
@@ -40,6 +49,85 @@ def test_build(self):
         assert len(workflow.__str__()) > 0
         assert workflow.rdf is not None
 
+    def test_construct_from_rdf_including_steps(self):
+        """
+        Construct FairWorkflow from RDF that includes detailed information
+        about steps.
+        """
+        rdf = self._get_rdf_test_resource('test_workflow_including_steps.trig')
+        uri = 'http://www.example.org/workflow1'
+        workflow = FairWorkflow.from_rdf(rdf, uri, fetch_steps=False)
+        valid_step_uris = [uri + '#step1', uri + '#step2', uri + '#step3']
+        steps = list(workflow)
+        assert len(steps) == 3
+        for step in steps:
+            step.validate()
+            assert step.uri in valid_step_uris
+        workflow.validate()
+
+    @mock.patch('fairworkflows.fairworkflow.FairWorkflow._fetch_step')
+    def test_construct_from_rdf_fetch_steps(self, mock_fetch_step):
+        """
+        Construct FairWorkflow from RDF that only includes URIs that point to
+        steps which we should fetch.
+        """
+        mock_fetch_step.return_value = self.step1
+        rdf = self._get_rdf_test_resource('test_workflow_excluding_steps.trig')
+        uri = 'http://www.example.org/workflow1'
+        workflow = FairWorkflow.from_rdf(rdf, uri, fetch_steps=True)
+        assert len(workflow._steps) == 1
+        assert mock_fetch_step.call_count == 1
+        assert list(workflow._steps.values())[0] == self.step1
+
+    def test_construct_from_rdf_do_not_fetch_steps(self):
+        """
+        Construct FairWorkflow from RDF that only includes URIs that point to
+        steps. We choose not to fetch these steps and have empty fair steps
+        pointing to the uri.
+        """
+        rdf = self._get_rdf_test_resource('test_workflow_excluding_steps.trig')
+        uri = 'http://www.example.org/workflow1'
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            workflow = FairWorkflow.from_rdf(rdf, uri, fetch_steps=False)
+            assert len(w) == 1
+            assert 'Could not get detailed information' in str(w[0].message)
+        assert len(workflow._steps) == 1
+
+    @mock.patch('fairworkflows.fairworkflow.FairWorkflow._fetch_step')
+    def test_construct_from_rdf_fetch_steps_fails(self, mock_fetch_step):
+        """
+        Construct FairWorkflow from RDF that only includes URIs that point to
+        steps. Fetching the step fails in this scenario.
+        """
+        mock_fetch_step.return_value = None
+        rdf = self._get_rdf_test_resource('test_workflow_excluding_steps.trig')
+        uri = 'http://www.example.org/workflow1'
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            workflow = FairWorkflow.from_rdf(rdf, uri, fetch_steps=True)
+            assert len(w) == 1
```

Comment:
I'm not entirely sure what this is doing. I suppose w ends up being a list of warnings but there can never be more than one in this case?

Reply:
(The ```assert len(w) == 1``` I mean)

Reply:
There should be exactly 1 warning being raised.

Reply:
I have added a message to the assert statements so that it is clearer.